pkg/sql/schemachanger/sctest/comparator.go

```diff
@@ -0,0 +1,283 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sctest
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/seqexpr"
+	"github.com/cockroachdb/cockroach/pkg/sql/parser"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scdeps/sctestutils"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/errors"
+	"github.com/lib/pq"
+	"github.com/stretchr/testify/require"
+)
+
+// This file contains common logic for performing comparator testing
+// between legacy and declarative schema changer.
+
+type StmtLineReader interface {
+	// HasNextLine returns true if there is more lines of SQL statements to process.
+	HasNextLine() bool
+
+	// NextLine retrieve the next line of SQL statements to process.
+	NextLine() string
+}
+
+// CompareLegacyAndDeclarative is the core logic for performing comparator
+// testing between legacy and declarative schema changer.
+// It reads sql statements (mostly DDLs), one by one, from `ss` and execute them
+// in a cluster using legacy schema changer and in another using declarative
+// schema changer. It asserts that, if the statement fails, they must fail with
+// a PG error with the same pg code. If the statement succeeded, all descriptors
+// in the cluster should end up in the same state.
+func CompareLegacyAndDeclarative(t *testing.T, ss StmtLineReader) {
+	ctx := context.Background()
+	var linesExecutedSoFar []string
+	legacyTSI, legacySQLDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	declarativeTSI, declarativeSQLDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer legacyTSI.Stopper().Stop(ctx)
```

Comment:
nit: This block is pretty difficult to read through IMO. Consider instead coupling the setup for each server type and separating them by a new line. You'll be able to see the symetric (or lack thereof) between the two and understand that two servers are being setup much easier.

```go
defer legacyTSI.Stopper().Stop(ctx)
legacyTDB := sqlutils.MakeSQLRunner(legacySQLDB)
legacyTDB.Exec(t, "SET use_declarative_schema_changer = off;")

defer declarativeTSI.Stopper().Stop(ctx)
declarativeTDB := sqlutils.MakeSQLRunner(declarativeSQLDB)
declarativeTDB.Exec(t, "SET use_declarative_schema_changer = on;")
```